### ISSUE TITLE
job launch fails with srun: [cn05] [[31923,0],1] selected pml ob1, but peer [[31923,0],0] on cn04 selected pml

### ISSUE THREAD

**Issue by bhendersonPlano** [open-mpi/ompi#12475]
I'm having an issue with a two node job launch when using srun and looking for some help.  A normal mpirun invoking mpirun from within an salloc works fine.

The system is a small number of dual socket nodes running RHEL 9.3 each with an Intel E810-C card (only port zero has a connection) and there are no other network cards in the system. The network connection is configured in an active-backup bond which I know is odd, but how our imaging tool likes things.  There is a single 100G switch and only one subnet.  /home is shared across the nodes via NFS.

Software is OpenMPI (5.0.3) with user built hwloc (2.10.0), pmix (5.0.2) and slurm (23.11.6).

Build options are:
```
./configure \
     --prefix=/home/brent/sys/openmpi/5.0.3 \
     --with-hwloc=/home/brent/sys/hwloc/2.10.0 \
     --with-slurm=/home/brent/sys/slurm/23.11.5 \
     --with-pmix=/home/brent/sys/pmix/5.0.2 \
     --disable-ipv6 \
     --enable-orterun-prefix-by-default
```
Slurm was also built with the same versions of hwloc and pmix.

-----------------------------

A simple test works fine with mpirun directly:
```
$ mpirun -n 2 -H cn03,cn04 ./hello_mpi.503
Hello from rank 0 on cn03
Hello from rank 1 on cn04
```

An mpirun from within an salloc works fine:
```
$ salloc -w cn03,cn04 --ntasks-per-node=1 mpirun ./hello_mpi.503 
salloc: Granted job allocation 136
salloc: Nodes cn[03-04] are ready for job
Hello from rank 0 on cn03
Hello from rank 1 on cn04
salloc: Relinquishing job allocation 136
```

However a srun does not work:
``` $ srun -w cn03,cn04 --ntasks-per-node=1 ./hello_mpi.503 
[cn04:545610] [[55359,0],1] selected pml ob1, but peer [[55359,0],0] on cn03 selected pml 

srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
slurmstepd: error: *** STEP 138.0 ON cn03 CANCELLED AT 2024-04-18T13:05:02 ***
srun: error: cn04: task 1: Exited with exit code 14
srun: Terminating StepId=138.0
srun: error: cn03: task 0: Killed
```

Note that on the error line, the second node is not showing what it did select - which seems odd.  It does not matter which two nodes I select, the first node (cn04 above) complains about the second node (cn03 above) not matching in the pml selection.

I added OMPI_MCA_btl_base_verbose=100 OMPI_MCA_pml_ob1_verbose=100 to my launch, cleaned up the output a little and then diffed the two.  No differences were shown up to the point where the error message comes out.  

```
$ diff good bad
26c26
< [cn04] btl:tcp: 0x904660: if bond0 kidx 6 cnt 0 addr 10.23.0.14 IPv4 bw 100000 lt 100
---
> [cn04] btl:tcp: 0x24e5830: if bond0 kidx 6 cnt 0 addr 10.23.0.14 IPv4 bw 100000 lt 100
57c57
< [cn03] btl:tcp: 0xc33660: if bond0 kidx 7 cnt 0 addr 10.23.0.13 IPv4 bw 100000 lt 100
---
> [cn03] btl:tcp: 0x1a69830: if bond0 kidx 7 cnt 0 addr 10.23.0.13 IPv4 bw 100000 lt 100
63,84c63,70
< [cn03] mca: bml: Using self btl for send to [[56142,1],0] on node cn03
< [cn04] mca: bml: Using self btl for send to [[56142,1],1] on node cn04
< Hello from rank 0 on cn03
< [cn04] mca: bml: Using tcp btl for send to [[56142,1],0] on node cn03
< [cn04] btl: tcp: attempting to connect() to [[56142,1],0] address 10.23.0.13 on port 1024
< [cn04] btl:tcp: would block, so allowing background progress
< [cn03] mca: bml: Using tcp btl for send to [[56142,1],1] on node cn04
< [cn03] btl: tcp: attempting to connect() to [[56142,1],1] address 10.23.0.14 on port 1024
< [cn03] btl:tcp: would block, so allowing background progress
< [cn03] btl:tcp: connect() to 10.23.0.14:1024 completed (complete_connect), sending connect ACK
< [cn03] btl:tcp: now connected to 10.23.0.14, process [[56142,1],1]
< [cn04] btl:tcp: connect() to 10.23.0.13:1024 completed (complete_connect), sending connect ACK
< [cn04] btl:tcp: now connected to 10.23.0.13, process [[56142,1],0]
< Hello from rank 1 on cn04
< [cn04] mca: base: close: component self closed
< [cn04] mca: base: close: unloading component self
< [cn04] mca: base: close: component tcp closed
< [cn04] mca: base: close: unloading component tcp
< [cn03] mca: base: close: component self closed
< [cn03] mca: base: close: unloading component self
< [cn03] mca: base: close: component tcp closed
< [cn03] mca: base: close: unloading component tcp
---
> [cn03] mca: bml: Using self btl for send to [[43367,0],0] on node cn03
> [cn04] [[43367,0],1] selected pml ob1, but peer [[43367,0],0] on cn03 selected pml 
> 
> slurmstepd: error: *** STEP 131.0 ON cn03 CANCELLED AT 2024-04-18T10:42:56 ***
> srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
> srun: error: cn04: task 1: Exited with exit code 14
> srun: Terminating StepId=131.0
> srun: error: cn03: task 0: Killed
```

I tried with OpenMPI 4.1.6 as well but ran into similar issues.  Figured it would be easier to debug with 5.x and it is where I would like to end up as well.  Any guidance on next steps would be appreciated.

**Comment by bhendersonPlano** [comment]
[bad_run_output.txt](https://github.com/open-mpi/ompi/files/15029514/bad_run_output.txt)

**Comment by rhc54** [comment]
Looks like everything ran okay, but then OMPI tries to re-establish connections during finalize - which fails because the messaging system already shutdown.

**Comment by bosilca** [comment]
Both processes initialize the BTLs, so it seems obvious they both selected the OB1 PML. But somehow that information is now properly exchaged between them. More precisely, it sounds as if two peers were in different modex collecting modes, one in direct and one in full, so they take different path in `mca_pml_base_pml_check_selected`. Or, and this would be the worst case scenario, the modex is somehow corrupted.

Let's make sure they both provide the same PML name. Put a breakpoint or a printf in `mca_pml_base_pml_selected` and dump the process name and the PML selected (or not).

**Comment by rhc54** [comment]
@bosilca If you look carefully at his output, you can see that we successfully completed `MPI_Init` and outputted his "hello" message. We then enter `MPI_Finalize`, the various btl components close - and then we attempt to re-enter a connection attempt, which fails. So it isn't the modex or anything in the initial channel that is failing - the question is: why are we attempting to reconnect after we finalize the BTLs?

**Comment by bosilca** [comment]
@rhc54 that is not correct, you seems to be confusing the "correct" and the "bad" output. If you take a look at the complete output posted you will notice the lack of application output, aka. the hello message is nowhere to be found, so we are not yet out of the `MPI_Init`.

**Comment by rhc54** [comment]
I see - very confusing the way that was presented. I stand corrected.

**Comment by bhendersonPlano** [comment]
Sorry for being confusing - will try a different approach next time.  

Looks like it is using the same pml on both processes.
```
$ srun --reservation=brent -N 2 -n 2 /home/brent/src/mpi/hello_mpi/hello_mpi.503
DEBUG: process 90471 is using pml ob1
DEBUG: process 89501 is using pml ob1
[cn10:89501] [[4419,0],1] selected pml ob1, but peer [[4419,0],0] on cn09 selected pml 

slurmstepd: error: *** STEP 166.0 ON cn09 CANCELLED AT 2024-04-18T16:34:52 ***
srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
srun: error: cn09: task 0: Killed
srun: Terminating StepId=166.0
srun: error: cn10: task 1: Exited with exit code 14
```

**Comment by rhc54** [comment]
FWIW: the code that exchanges the PML name, when combined with the `OPAL_MODEX_foo`macros to send/recv the data, probably isn't written as it should be. It treats strings as byte objects, which means you lose all the type checking PMIx provides. It would be better if one actually "put" a string, and then received a string, to ensure that things are properly matched and the strings terminated.

Given that `mpirun` is working correctly and you are getting identical results for both OMPI v4.1 and v5.0, I'm guessing it is something in the handoff with Slurm that is causing the problem. I note that you didn't specify `-mpi=pmix` on your `srun` cmd line - but I'm guessing you must have that locked in the Slurm config file? We know there are some issues with Slurm 23.11 (though the ones we are currently working thru aren't related to this), so it could be another problem in there as well. Have you filed an issue with them?

**Comment by bhendersonPlano** [comment]
Yep, we have 'MpiDefault=pmix' in the slurm.conf file so I didn't need it
on the srun command line.  Good suggestion about the slurm 23.11.x build.
I may drop back a major revision and see if things 'just work' before
opening an issue with the schedmd folks.  Thanks for the help.

**Comment by bhendersonPlano** [comment]
Thanks Ralph - I backed down to slurm 23.02.7 (with the same hwloc and
pmix) and saw the same issue.

My next experiment is to back down PMIx from 5.0.2 to 4.2.9.  Next on the
list is the PMIX_MCA_gds=hash setting and then patch above with the
pml_base_verbose.

I'll keep you posted on what I find out

**Comment by rhc54** [comment]
Thanks! Really appreciate the help in debugging this problem.

**Comment by rhc54** [comment]
Hmmm...sounds like we do have a problem in the shared memory component in PMIx. Rats - it's the hardest to chase down, and hardest problems to reproduce (I still cannot reproduce this problem locally). Trying out the patch would be appreciated - it would be good to know what data type the returned value was. Would tell us if the stored object was bad, or we are just missing the string.

FWIW: with Slurm using PMIx 4.2.9 and OMPI on 5.0.2, the shmem component will be disabled as the two versions shared memory implementations are incompatible. So it effectively forces the "gds=hash" setting.

Thanks again!

**Comment by bhendersonPlano** [comment]
looks like your suspicion was correct on receiving a non-string object:
```
[cn04] check:select: got a non-string (PMIX_BYTE_OBJECT) pml from process [[35399,0],0]
```
Full file here - [bad_run_output_patched.txt](https://github.com/open-mpi/ompi/files/15043395/bad_run_output_patched.txt)

let me know if there are any other runs you want/need for debugging purposes.

**Comment by rhc54** [comment]
If you wouldn't mind, perhaps you could apply this patch to the PMIx v5.0.2 code:
```diff
diff --git a/src/mca/gds/shmem2/gds_shmem2.c b/src/mca/gds/shmem2/gds_shmem2.c
index 6f7f2ea3..11b48f39 100644
--- a/src/mca/gds/shmem2/gds_shmem2.c
+++ b/src/mca/gds/shmem2/gds_shmem2.c
@@ -1860,6 +1860,8 @@ client_update_global_keyindex_if_necessary(
     pmix_keyindex_t tmpindex;
     bool found;

+    return PMIX_SUCCESS;
+
     // Do we need to update?
     pmix_gds_shmem2_job_t *job;
     rc = pmix_gds_shmem2_get_job_tracker(nspace_name, true, &job);
```
It would be interesting to see if that is the problem. If so, it means that we are losing track of the key-to-index translation, and so we are returning the wrong value when asked for the PML key. The patch should be okay as long as both Slurm and OMPI are using the same PMIx code.

**Comment by bhendersonPlano** [comment]
I applied the patch and rebuilt PMIx.  Next, I rebuilt slurm and openmpi pointing to that new build.  Good news is that everything worked.  I kind of wish I had wrapped that 'just return success' with an env variable so I could turn it on and off easily.  If I get some time later today maybe I'll do that, but for now - your patch above to the PMIx 5.0.2 tree worked great!

**Comment by rhc54** [comment]
Ah, excellent - many thanks! I think I know what is going on, but can take that up over on the PMIx ticket. Thanks again for all the help.

**Comment by vasvir** [comment]
Hi,

I believe I have been bitten by this. Debian unstable openmpi 4.1.6-13.3.

- Is there a timeframe for the fix?
- Another bug to watch? I only found this.
- Patches to apply / test?
    - https://github.com/open-mpi/ompi/issues/12475#issuecomment-2066637826
    - https://github.com/open-mpi/ompi/issues/12475#issuecomment-2066983163
    - both of the above?

**Comment by rhc54** [comment]
What version of PMIx are you using? Note that this reported issue has nothing to do with OMPI v4 itself (that release doesn't include a copy of PMIx).

**Comment by vasvir** [comment]
libpmix2t64:amd64 5.0.2-3

I have to say:
* I upgraded openmpi to 5.0.3 (experimental) and the issue went away (for now)
* It really looks like a memory corruption to me because the error message says that one "node selected pml ob1 but the other node selected pml <C0A8>" which is invalid UTF-8 garbage...

**Comment by rhc54** [comment]
No, it isn't memory corruption. OMPI's v5 branch currently contains a newer version of PMIx that contains the fix but hasn't been released yet. PMIx v5.0.3 Probably gets released later this week.

**Comment by lrbison** [comment]
I ran into this problem again today and I thought I would spell out the workaround explicitly for myself and others:

```
export PMIX_MCA_gds=^shmem2
```

Then run your mpi program as before.

PMIx v5.0.3 as been released, which I understand fixes this issue, but setting the mca variable is a quick workaround.
